src/plugins/kibana_react/public/saved_objects/saved_object_finder.tsx

```diff
@@ -0,0 +1,520 @@
+
+import { SavedObjectAttributes } from '../../../../core/server';
+import { SimpleSavedObject, CoreStart } from '../../../../core/public';
+
+// TODO the typings for EuiListGroup are incorrect - maxWidth is missing. This can be removed when the types are adjusted
+const FixedEuiListGroup = (EuiListGroup as any) as React.FunctionComponent<
+  CommonProps & { maxWidth: boolean }
+>;
+
+// TODO the typings for EuiContextMenuPanel are incorrect - watchedItemProps is missing. This can be removed when the types are adjusted
+const FixedEuiContextMenuPanel = (EuiContextMenuPanel as any) as React.FunctionComponent<
+  EuiContextMenuPanelProps & { watchedItemProps: string[] }
+>;
+
+export interface SavedObjectMetaData<T extends SavedObjectAttributes> {
+  type: string;
+  name: string;
+  getIconForSavedObject(savedObject: SimpleSavedObject<T>): IconType;
+  getTooltipForSavedObject?(savedObject: SimpleSavedObject<T>): string;
+  showSavedObject?(savedObject: SimpleSavedObject<T>): boolean;
+}
+
+interface SavedObjectFinderState {
+  items: Array<{
+    title: string | null;
+    id: SimpleSavedObject<SavedObjectAttributes>['id'];
+    type: SimpleSavedObject<SavedObjectAttributes>['type'];
+    savedObject: SimpleSavedObject<SavedObjectAttributes>;
+  }>;
+  query: string;
+  isFetchingItems: boolean;
+  page: number;
+  perPage: number;
+  sortDirection?: Direction;
+  sortOpen: boolean;
+  filterOpen: boolean;
+  filteredTypes: string[];
+}
+
+interface BaseSavedObjectFinder {
+  onChoose?: (
+    id: SimpleSavedObject<SavedObjectAttributes>['id'],
+    type: SimpleSavedObject<SavedObjectAttributes>['type'],
+    name: string
+  ) => void;
+  noItemsMessage?: React.ReactNode;
+  savedObjectMetaData: Array<SavedObjectMetaData<SavedObjectAttributes>>;
+  showFilter?: boolean;
+}
+
+interface SavedObjectFinderFixedPage extends BaseSavedObjectFinder {
+  initialPageSize?: undefined;
+  fixedPageSize: number;
+}
+
+interface SavedObjectFinderInitialPageSize extends BaseSavedObjectFinder {
+  initialPageSize?: 5 | 10 | 15 | 25;
+  fixedPageSize?: undefined;
+}
+type SavedObjectFinderProps = {
+  savedObjects: CoreStart['savedObjects'];
+  uiSettings: CoreStart['uiSettings'];
+} & (SavedObjectFinderFixedPage | SavedObjectFinderInitialPageSize);
+
+class SavedObjectFinder extends React.Component<SavedObjectFinderProps, SavedObjectFinderState> {
+  public static propTypes = {
+    onChoose: PropTypes.func,
+    noItemsMessage: PropTypes.node,
+    savedObjectMetaData: PropTypes.array.isRequired,
+    initialPageSize: PropTypes.oneOf([5, 10, 15, 25]),
+    fixedPageSize: PropTypes.number,
+    showFilter: PropTypes.bool,
+  };
+
+  private isComponentMounted: boolean = false;
+
+  private debouncedFetch = _.debounce(async (query: string) => {
+    const metaDataMap = this.getSavedObjectMetaDataMap();
+
+    const perPage = this.props.uiSettings.get('savedObjects:listingLimit');
+    const resp = await this.props.savedObjects.client.find({
```

Comment:
Hasn't changed since the old component, but shouldn't we add error handling for example for network failures?

Reply:
Yes, this whole component seems to be too big and implemented without handling errors or async behaviour. Maybe there could be a separate task to refactor it all.

Reply:
Created an issue: https://github.com/elastic/kibana/issues/44095